Compress and cache cluster state during validate join request

amkhar opened the pull request:
<!--  Thanks for sending a pull request, here are some tips:

1. If this is a fix for an undisclosed security vulnerability, please STOP. All security vulnerability reporting and fixes should be done as per our security policy https://github.com/opensearch-project/OpenSearch/security/policy
2. If this is your first time, please read our contributor guidelines: https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md and developer guide https://github.com/opensearch-project/OpenSearch/blob/main/DEVELOPER_GUIDE.md
3. Ensure you have added or ran the appropriate tests for your PR: https://github.com/opensearch-project/OpenSearch/blob/main/TESTING.md
-->
### Description
ValidateJoinRequest flow sends the full cluster state to every node by writing it on the stream separately for each node. We can improve this by compressing the cluster state first and storing it in a cache so the same serialized data is written over and over on different tcp channels instead of writing it fresh. 
We're seeing good amount of improvement if we refresh the serialzed state cache every 30s and use similar compression technique used during cluster state publishing (in PublicationTransportHandler). Just with compression, gain is less but we can see it in flamegraph. With caching & compression, gain is huge. Overall memory allocations for sendValidateJoinRequest flow is reduced to 1/7th

ToDo : Unit tests and more fine tuning to be done. [DONE]

Pending Action Items 
* Compare CPU profile during node drops - DONE. Overall 30% less CPU utilization for sendValidateJoinRequest flow.
* Run another load test where instead of index metadata other fields are also high in size in cluster state. 

### Issues Resolved
https://github.com/opensearch-project/OpenSearch/issues/5100

### Check List
- [x] New functionality includes testing.
  - [x] All tests pass
- [x] New functionality has been documented.
  - [x] New functionality has javadoc added
- [x] Commits are signed per the DCO using --signoff
- [x] Commit changes are listed out in CHANGELOG.md file (See: [Changelog](../blob/main/CONTRIBUTING.md#changelog))

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

shwetathareja commented on server/src/main/java/org/opensearch/cluster/CompressionHelper.java:
```diff
@@ -0,0 +1,49 @@
+     * It'll always use compression before writing on a newly created output stream.
+     * @param writer Object which is going to write the content
+     * @param nodeVersion version of cluster node
+     * @param streamBooleanFlag flag used at receiver end to make intelligent decisions. For example, ClusterState
```
This flag looks specific to ClusterState and doesnt seem generic and node version

shwetathareja replied:
keep the method name as serializeClusterState and you can use this method from  for both full and diff cluster state. 

This class is compressing and serializing both. May be you name it specific to ClusterState

amkhar replied:
These flags or fields we can write on stream even multiple times, then in same manner you read on receiver end write your logic around the values. My thought is to create a helper class which in future can be used by other communications also if anyone wants to use compression for a change or just want to use BytesTransportRequest (in that case another helper method can be added without compression). That's why I want to keep this class generic. 

My proposal : Let's keep the class generic, as method is specifically writing the bytes on stream specific for cluster state use case, I'll mark method name as serializeClusterState.

Also will move it to opensearch.common.compress so other code base can easily use it. Let me know if you're aware of a better sub package.

shwetathareja commented on server/src/main/java/org/opensearch/cluster/coordination/PublicationTransportHandler.java:
```diff
@@ -309,15 +310,15 @@ void buildDiffAndSerializeStates() {
                     if (sendFullVersion || previousState.nodes().nodeExists(node) == false) {
                         if (serializedStates.containsKey(node.getVersion()) == false) {
-                            serializedStates.put(node.getVersion(), serializeFullClusterState(newState, node.getVersion()));
+                            serializedStates.put(node.getVersion(), CompressionHelper.serializedWrite(newState, node.getVersion(), true));
```
you will clean up the ```serializeFullClusterState``` and ```serializeDiffClusterState``` from this class

amkhar replied:
Yes, this is just draft PR.

shwetathareja commented on server/src/main/java/org/opensearch/common/settings/ClusterSettings.java:
```diff
@@ -638,7 +638,8 @@ public void apply(Settings value, Settings current, Settings previous) {
                 // Settings related to Searchable Snapshots
-                Node.NODE_SEARCH_CACHE_SIZE_SETTING
+                Node.NODE_SEARCH_CACHE_SIZE_SETTING,
+                JoinHelper.CLUSTER_MANAGER_JOIN_STATE_REFRESH_INTERVAL
```
CLUSTER_MANAGER_VALIDATE_JOIN_CACHE_INTERVAL

amkhar replied:
Ack.

shwetathareja commented on server/src/main/java/org/opensearch/cluster/CompressionHelper.java:
```diff
@@ -0,0 +1,49 @@
+     * @return reference to serialized bytes
+     * @throws IOException if writing on the compressed stream is failed.
+     */
+    public static BytesReference serializedWrite(Writeable writer, Version nodeVersion, boolean streamBooleanFlag) throws IOException {
```
dont create static methods, it makes testing and mocking difficult.

amkhar replied:
Mockito can now mock static method easily and we use 5.x version of Mockito. All utility classes where no instance related information is required, we don't actually create the object of those classes and simply use static methods. Footprint is also smaller for static methods. 
I'll still see if there is any scope of keeping instance variables inside the helper class, if not I prefer to keep methods static.

shwetathareja commented on server/src/main/java/org/opensearch/cluster/coordination/JoinHelper.java:
```diff
@@ -206,24 +230,55 @@ public ClusterTasksResult<JoinTaskExecutor.Task> execute(ClusterState currentSta
+        );
+    }
+
+    private void handleValidateJoinRequest(
+        Supplier<ClusterState> currentStateSupplier,
+        Collection<BiConsumer<DiscoveryNode, ClusterState>> joinValidators,
+        BytesTransportRequest request
+    ) throws IOException {
+        final Compressor compressor = CompressorFactory.compressor(request.bytes());
+        StreamInput in = request.bytes().streamInput();
+        final ClusterState incomingState;
+        try {
+            if (compressor != null) {
+                in = new InputStreamStreamInput(compressor.threadLocalInputStream(in));
+            }
+            in = new NamedWriteableAwareStreamInput(in, namedWriteableRegistry);
+            in.setVersion(request.version());
+            // If true we received full cluster state - otherwise diffs
+            if (in.readBoolean()) {
+                // Close early to release resources used by the de-compression as early as possible
+                try (StreamInput input = in) {
+                    incomingState = ClusterState.readFrom(input, transportService.getLocalNode());
+                } catch (Exception e) {
+                    logger.warn("unexpected error while deserializing an incoming cluster state", e);
```
This would be common code with cluster state publishing deserialization. Can we move this also to common place?

shwetathareja commented on server/src/main/java/org/opensearch/cluster/coordination/JoinHelper.java:
```diff
@@ -407,12 +462,32 @@ public String executor() {
+                try {
+                    // Re-getting current cluster state for validate join request
+                    bytes = CompressionHelper.serializedWrite(state, node.getVersion(), true);
+                    serializedStates.put(node.getVersion(), bytes);
```
```serializedStates``` is just HashMap which is not threadSafe and multiple send ValidateJoin can overwrite the map. Also, lets say multiple threads enter this code path either when bytes is null or cache expired. Each one will end up calculating the bytes and populating the HashMap. There is a Cache class - ```org.opensearch.common.cache.Cache``` you can use that for auto expiry and you can use computeIfAbsent() to ensure only one thread serializes at any time.

amkhar replied:
Ack.

shwetathareja commented on server/src/main/java/org/opensearch/cluster/coordination/JoinHelper.java:
```diff
@@ -206,24 +230,55 @@ public ClusterTasksResult<JoinTaskExecutor.Task> execute(ClusterState currentSta
+        );
+    }
+
+    private void handleValidateJoinRequest(
```
This logic would need to account for the input/ output stream version as there could be mixed mode clusters
1. Leader sends old cluster state (uncompressed) while node n1 expects uncompressed state -- No issue
2. Leader sends old cluster state (uncompressed) while node n1 expects compressed state  --- It would fail
3. Leader sends compressed state while node n1 expects compressed state  -- No Issue
4. Leader sends compressed state while node n1 expects uncompressed state.   --- It would fail (though this case, OpenSearch doesn't guarantee)

Code needs to handle use case for 2 & 4

amkhar replied:
Adding a different action name for sending compressed state as well as adding handler separately. Thanks I missed this.

amkhar commented:
Verified org.opensearch.remotestore.SegmentReplicationRemoteStoreIT.testReplicaHasDiffFilesThanPrimary test is running fine on local, it should be a flaky test. Pushed another commit to see. Else will create a bug for this IT.

sandeshkr419 commented on server/src/main/java/org/opensearch/cluster/coordination/PublicationTransportHandler.java:
```diff
@@ -309,15 +255,15 @@ void buildDiffAndSerializeStates() {
                 try {
                     if (sendFullVersion || previousState.nodes().nodeExists(node) == false) {
```
Should we refactor `sendFullVersion` to `isFullVersion` or something along similar lines and pass this value when invoking `serializeClusterState(.., .., isFullVersion)` instead of hardcoding true/false. Will make the code more readable if similar changes can be done at other invocation of this method.

amkhar replied:
In one of the conditions, if `sendFullVersion` is true, then also there is a case where we'll send diff. So, not planning to rely on this variable. Also during exception handling while sending cluster state, we need to send full cluster state and this variable doesn't contribute in that flow. So, eventually we'll anyway need to pass true/false manually.

Keeping it as it is.

sandeshkr419 reviewed (changes requested):
Some minor refactoring comments on code.

shwetathareja commented on server/src/main/java/org/opensearch/cluster/coordination/JoinHelper.java:
```diff
@@ -407,12 +475,27 @@ public String executor() {
+                    key -> CompressionHelper.serializeClusterState(state, node, true)
+                );
+            }
+            final BytesTransportRequest request = new BytesTransportRequest(bytes, node.getVersion());
+            transportService.sendRequest(
+                node,
+                VALIDATE_COMPRESSED_JOIN_ACTION_NAME,
+                request,
+                new ActionListenerResponseHandler<>(listener, i -> Empty.INSTANCE, ThreadPool.Names.GENERIC)
```
ClusterManager should choose between the regular ValidateJoinRequest or compressed one based on the node version (you can get this by checking stream version). If it always sends compressed one then data nodes on older version wont be able to parse it.

amkhar replied:
Agreed, thanks for catching this. I'll add a check for 3.0 version to send compressed state.

amkhar replied:
Added a check for 2.8 considering other ongoing PRs.

andrross replied:
Not sure what you mean by "other ongoing PRs"? I think the right way to do this is to add an "on or after 3.0" check, then backport this and change it to "on or after 2.8" in the 2.x backport PR, then immediately open another PR on main to change this to be "on or after 2.8" once the backport is merged. It's a bit cumbersome but it does guarantee that all tests will pass in every PR. I believe the current PR revision will attempt to send the compressed version to 2.8, which won't know how to parse it.

shwetathareja reviewed (commented):
Thanks @amkhar, Posted initial comment, I am still reviewing the changes.

shwetathareja commented on server/src/main/java/org/opensearch/common/compress/CompressionHelper.java:
```diff
@@ -0,0 +1,99 @@
+/**
+ * A helper class to utilize the compressed stream.
+ */
+public class CompressionHelper {
```
I would still recommend renaming this class ClusterStateUtils as it has method for compression and serialization both.

shwetathareja replied:
This class can be made generic in future if there are really other objects looking for similar functionality.

shwetathareja replied:
Mark this class final

shwetathareja commented on server/src/main/java/org/opensearch/common/settings/ClusterSettings.java:
```diff
@@ -649,7 +649,10 @@ public void apply(Settings value, Settings current, Settings previous) {
+                RemoteRefreshSegmentPressureSettings.UPLOAD_TIME_MOVING_AVERAGE_WINDOW_SIZE,
+
+                // Settings related to node join
+                JoinHelper.CLUSTER_MANAGER_VALIDATE_JOIN_CACHE_INTERVAL
```
[minor]: there is another JoinHelper setting above L565, keep it along with that

amkhar replied:
Ack.

Bukhtawar replied:
wondering why would you want to keep it time based, wouldn't you want to invalidate this on a newer version? How does time help? Is this about bounded consistency?

amkhar replied:
@Bukhtawar  explained in detail on this thread https://github.com/opensearch-project/OpenSearch/pull/7321#discussion_r1200219753

amkhar replied:
Update : removing the setting.

shwetathareja commented on server/src/test/java/org/opensearch/cluster/coordination/NodeJoinTests.java:
```diff
@@ -213,7 +213,7 @@ protected void onSendRequest(long requestId, String action, TransportRequest req
                         requestId,
                         new TransportService.HandshakeResponse(destination, initialState.getClusterName(), destination.getVersion())
                     );
-                } else if (action.equals(JoinHelper.VALIDATE_JOIN_ACTION_NAME)) {
```
Please update and keep tests for both actions.

amkhar replied:
Keeping it as it is and adding VALIDATE_COMPRESSED_JOIN_ACTION_NAME case also.

amkhar commented on server/src/main/java/org/opensearch/cluster/coordination/JoinHelper.java:
```diff
@@ -98,6 +107,7 @@ public class JoinHelper {
 
     public static final String JOIN_ACTION_NAME = "internal:cluster/coordination/join";
     public static final String VALIDATE_JOIN_ACTION_NAME = "internal:cluster/coordination/join/validate";
+    public static final String VALIDATE_COMPRESSED_JOIN_ACTION_NAME = "internal:cluster/coordination/join" + "/validate_compressed";
```
Ack.

amkhar commented on server/src/main/java/org/opensearch/cluster/coordination/JoinHelper.java:
```diff
@@ -208,22 +237,61 @@ public ClusterTasksResult<JoinTaskExecutor.Task> execute(ClusterState currentSta
+        joinValidators.forEach(action -> action.accept(transportService.getLocalNode(), incomingState));
+    }
+
+    private void handleValidateJoinRequest(
```
Ack.

shwetathareja reviewed (commented):
@amkhar : 
1. Please add tests for both transport action ensuring based on node version, right handler transport action is sent.
2. Test for CompressionHelper class
3.  Test for handling ValidateJoin both versions on receiving node

amkhar commented on server/src/main/java/org/opensearch/common/compress/CompressionHelper.java:
```diff
@@ -0,0 +1,99 @@
+    }
+
+    public static Diff<ClusterState> deserializeClusterStateDiff(StreamInput in, DiscoveryNode localNode) throws IOException {
+        final Diff<ClusterState> diff;
```
Ack.

amkhar commented on server/src/main/java/org/opensearch/common/compress/CompressionHelper.java:
```diff
@@ -0,0 +1,99 @@
+    public static StreamInput decompressClusterState(BytesTransportRequest request, NamedWriteableRegistry namedWriteableRegistry)
+        throws IOException {
+        final Compressor compressor = CompressorFactory.compressor(request.bytes());
+        StreamInput in = request.bytes().streamInput();
```
Normally StreamOutput and StreamInput are used for node-to-node communications. We write on StreamOutput then read from StreamInput.

Adding javadoc for clarity.

amkhar commented on server/src/main/java/org/opensearch/cluster/coordination/JoinHelper.java:
```diff
@@ -208,22 +237,61 @@ public ClusterTasksResult<JoinTaskExecutor.Task> execute(ClusterState currentSta
+        Collection<BiConsumer<DiscoveryNode, ClusterState>> joinValidators,
+        BytesTransportRequest request
+    ) throws IOException {
+        StreamInput in = CompressionHelper.decompressClusterState(request, namedWriteableRegistry);
```
Ack.

andrross commented on server/src/main/java/org/opensearch/cluster/coordination/JoinHelper.java:
```diff
@@ -122,6 +132,16 @@ public class JoinHelper {
+    private Cache<Version, BytesReference> serializedStates;
+    public static final Setting<TimeValue> CLUSTER_MANAGER_VALIDATE_JOIN_CACHE_INTERVAL = Setting.timeSetting(
+        "cluster_manager.validate_join.cache_interval",
+        TimeValue.timeValueMillis(30000),
+        TimeValue.timeValueMillis(0),
+        TimeValue.timeValueMillis(60000),
```
Another nitpick...probably more readable to use `timeValueSeconds`

amkhar replied:
Ack.

andrross commented on server/src/main/java/org/opensearch/cluster/coordination/JoinHelper.java:
```diff
@@ -208,22 +237,61 @@ public ClusterTasksResult<JoinTaskExecutor.Task> execute(ClusterState currentSta
+        BytesTransportRequest request
+    ) throws IOException {
+        StreamInput in = CompressionHelper.decompressClusterState(request, namedWriteableRegistry);
+        ClusterState incomingState;
```
Can you move this variable definition down to the assignment below?

amkhar replied:
It's already done in my workspace :) as part of other changes, just waiting to push a commit. 👍🏼

andrross reviewed (commented):
Just to make sure I understand, this will result in always sending the compressed payload when connected to a compatible node? And the uncompressed version is only sent to handle backward compatibility with previous versions of the code.

andrross commented on server/src/main/java/org/opensearch/cluster/coordination/JoinHelper.java:
```diff
@@ -407,12 +475,27 @@ public String executor() {
+                bytes = CompressionHelper.serializeClusterState(state, node, true);
+            } else {
+                bytes = serializedStates.computeIfAbsent(
+                    node.getVersion(),
```
I asked the question [here](https://github.com/opensearch-project/OpenSearch/issues/5100#issuecomment-1551713787) as well, but can this be a composite key with `state.version()` so that we always recompute the state if a newer state exists and never send stale data? If so, we could probably get rid of the configurable interval setting, and possibly replace the cache object with simpler memoization.

amkhar replied:
Responded [here](https://github.com/opensearch-project/OpenSearch/issues/5100#issuecomment-1552738053) well. Just going with state-version based caching won't give enough benefit for customers. At least for the scenario we're trying to solve, evicting cache with every node-join update would be non-deterministic for the benefits. We may see some improvements sometimes but not all the time. 

A mixed approach of both state-version based caching and time based would be more helpful. We can keep the default behaviour of no caching in 2.x but change it for 3.0 with 30s cache to make sure customers see significant improvement during sudden multi node joins.

Let's close the approach on https://github.com/opensearch-project/OpenSearch/issues/5100 conversation, based on that I'll push next commit.

Thanks @andrross for all the useful comments 👍🏼

Bukhtawar replied:
@amkhar It isn't very clear as to what are the trade-offs of caching? Can we run into nodes with incompatible cluster state join the cluster? Whats the rationale behind 30s meaning what do we lose if we make this 30 mins? I would second @andrross's recommendation on memoization is the idea is that diffs would be minimal but frequent to invalidate caches on cluster manager.
IMO the settings makes me feel we are building a semantic that can be misused, can leak to breakages of cluster state guarantees with consistency in future.

amkhar replied:
Thanks @Bukhtawar for review.
It would be easy to close the approach on https://github.com/opensearch-project/OpenSearch/issues/5100 conversation as it has all the background and experiment results which will answer some of the questions.

>Can we run into nodes with incompatible cluster state join the cluster?

Cluster state is sent by Cluster Manager, so if we send cluster state of 10s ago, data node is supposed to run validations on the index metadata of incoming cluster state with its node version to make sure correct version of data node will send a success in return of that join validation request(from cluster manager). So, node will run validations on that data of 10s ago. Now Cluster Manager will anyway run those validations as part of node-join task through TaskBatcher and fail the join if any validation fails. Only thing which we need to keep in mind is that, if any of the discovery plugin has some extra validation, that will work on 10s older state. Today, **we don't have any validators from ec2, azure or gce**.

>Whats the rationale behind 30s meaning what do we lose if we make this 30 mins?

Index created during this interval(30s or 30min) won't be part of the index metadata of cluster state which is being sent for validation to data nodes.

> I would second @andrross's recommendation on memoization is the idea is that diffs would be minimal but frequent to invalidate caches on cluster manager.

Sure, I'm running some tests with just state-version based caching where we invalidate the cache if cluster state is updated. I'll share the results so we can collaboratively take a decision.

amkhar replied:
>Whats the rationale behind 30s

I choose this interval after running experiments on other values like 10s also. So most benefit was in 30s and it takes around 1 minute for 100 nodes to join the cluster back so whatever best caching we can provide in that interval is going to reflect in allocation gains.

amkhar replied:
Update : using state's version based caching only after seeing results https://github.com/opensearch-project/OpenSearch/issues/5100#issuecomment-1557087555

amkhar commented on server/src/main/java/org/opensearch/cluster/coordination/JoinHelper.java:
```diff
@@ -122,6 +132,16 @@ public class JoinHelper {
     private final Consumer<Boolean> nodeCommissioned;
+    private final NamedWriteableRegistry namedWriteableRegistry;
+    private Cache<Version, BytesReference> serializedStates;
+    public static final Setting<TimeValue> CLUSTER_MANAGER_VALIDATE_JOIN_CACHE_INTERVAL = Setting.timeSetting(
```
Sure, will move it up and add comment for explanation.

Bukhtawar commented on server/src/main/java/org/opensearch/common/compress/CompressionHelper.java:
```diff
@@ -0,0 +1,99 @@
+    public static BytesReference serializeClusterState(Writeable writer, DiscoveryNode node, boolean isFullClusterState)
+        throws IOException {
+        try {
+            final BytesStreamOutput bStream = new BytesStreamOutput();
```
Ensure you handle closing streams

amkhar replied:
Yes, it's the same code which is already being used in PublicationTransportHandler. We can't close stream until it's written on the transport layer. Here only ByteReference is taken as output. So, later in code flow, when it's written on transport channel using the same byte reference, it'll be released.

Bukhtawar commented on server/src/main/java/org/opensearch/common/compress/CompressionHelper.java:
```diff
@@ -0,0 +1,99 @@
+        final ClusterState incomingState;
+        try (StreamInput input = in) {
+            incomingState = ClusterState.readFrom(input, localNode);
+        } catch (Exception e) {
+            logger.warn("unexpected error while deserializing an incoming cluster state", e);
+            throw e;
+        }
```
whats the point catching and re-throwing the same exception, this might lead to logging redundant stack traces

andrross replied:
Agreed. Log-and-throw is an anti-pattern because the thing that ultimately handles this exception has the most context of what to do (and usually logs the exception anyway, leading to duplicate logging and stack traces).

amkhar commented:
@shwetathareja 

>Please add tests for both transport action ensuring based on node version, right handler transport action is sent.

Adding testValidateJoinSentWithCorrectActionForVersions.

>Test for ClusterStateUtils class

Added all necessary functional tests in JoinHelperTests, can't write a separate test class for static methods.

>Test for handling ValidateJoin both versions on receiving node

Updated testJoinValidationRejectsMismatchedClusterUUID for compressed validate join flow.

andrross commented on server/src/main/java/org/opensearch/cluster/coordination/JoinHelper.java:
```diff
@@ -122,6 +131,8 @@ public class JoinHelper {
     private final Supplier<JoinTaskExecutor> joinTaskExecutorGenerator;
     private final Consumer<Boolean> nodeCommissioned;
+    private final NamedWriteableRegistry namedWriteableRegistry;
+    private final Cache<Long, BytesReference> serializedStates;
```
Do we need to use a `Cache` here? Would a `HashMap` or `ConcurrentHashMap` be sufficient?

amkhar replied:
Yes! Normal ConcurrentHashMap is good enough. Changing it back to CHM.

andrross commented on server/src/main/java/org/opensearch/cluster/coordination/JoinHelper.java:
```diff
@@ -407,12 +462,35 @@ public String executor() {
+                    new ActionListenerResponseHandler<>(listener, i -> Empty.INSTANCE, ThreadPool.Names.GENERIC)
+                );
+            } catch (Exception e) {
+                logger.warn(() -> new ParameterizedMessage("error sending cluster state to {}", node), e);
+                listener.onFailure(e);
```
This is potentially the same thing as log-and-throw. What does the listener do with this failure? Will we get a duplicate stack trace in the logs?

amkhar replied:
Listener.onFailure is necessary.
This listener is passed via Join request, it's basically the transport joinCallBack listener. So, if anything wrong happens, data node should be notified that its join request has failed. 

The flow : 
* cluster manager creates a transportJoinCallback and sets it as part of join handler
* data-node sends a join request to cluster manager
* cluster manager after receiving the join request, tries to connect to the node
* Once connection is successful, cluster manager send the validateJoinRequest
* If anything fails in any of these flow, that transport join call back .failure method needs to be invoked for proper communication back to data node

In case of any other exception apart from transport call of sendValidateJoinRequest, we need to trigger the joinCallBack listener's failure code.

JoinCallback's flow will log that `failed to validate incoming join request from node [{}]`
and validate join(this flow) is logging `error sending cluster state to {node}` --> this log line helps people understand the cluster manager couldn't send cluster state to data node that's why validation of incoming join request from data node failed. Without this log, people who're not aware of code path won't understand why incoming join request from node was not validated.

Listener we need trigger, but log we can skip. Once we skip log, we'll only get **failed to validate incoming join request from node [{}]** in logs.

Just because it's a different transport call from cluster manager to data node, I felt it would be good to have a log line for the failure. Let me know what do you think.

andrross replied:
Since the listener logs the stack trace, maybe skip that here? Agreed that "error sending cluster state to {node}" can be helpful context but probably not necessary to print the exception stack trace twice.

amkhar replied:
Sure, will print just the log not the trace 👍🏼

amkhar commented on server/src/main/java/org/opensearch/cluster/coordination/ClusterStateUtils.java:
```diff
@@ -0,0 +1,96 @@
+import java.io.IOException;
+
+/**
+ * A helper class to utilize the compressed stream.
```
Ack.

shwetathareja commented on server/src/main/java/org/opensearch/cluster/coordination/ClusterStateUtils.java:
```diff
@@ -0,0 +1,96 @@
+/**
+ * A helper class to utilize the compressed stream.
+ */
+public final class ClusterStateUtils {
```
Add UTs for this class as well.

amkhar replied:
Ack.
Though, tests are added in JoinHelperTests class for testing the failures(`testJoinValidationFailsOnDecompressionFailure`, `testSendCompressedValidateJoinFailOnSerializeFailure`) of this class's methods, I'll still create a separate test class for this.

amkhar commented on server/src/internalClusterTest/java/org/opensearch/cluster/coordination/ZenDiscoveryIT.java:
```diff
@@ -117,7 +117,6 @@ public void testHandleNodeJoin_incompatibleClusterState() throws InterruptedExce
 
         final CompletableFuture<Throwable> future = new CompletableFuture<>();
         DiscoveryNode node = state.nodes().getLocalNode();
-
```
Removing it.

andrross commented on server/src/main/java/org/opensearch/cluster/coordination/JoinHelper.java:
```diff
@@ -407,12 +459,38 @@ public String executor() {
+            );
+        } else {
+            try {
+                if (!serializedStates.containsKey(state.version())) {
+                    serializedStates.clear();
+                }
+                BytesReference bytes = serializedStates.computeIfAbsent(state.version(), version -> {
+                    try {
+                        return ClusterStateUtils.serializeClusterState(state, node, true);
+                    } catch (IOException e) {
+                        // mandatory as ConcurrentHashMap doesn't rethrow IOException.
+                        throw new RuntimeException(e);
+                    }
+                });
```
Given that the map will only ever hold a single entry, what do you think about using an AtomicReference instead? Something like this:

```java
final AtomicReference<Tuple<Long, BytesReference>> serializedState = new AtomicReference<>();

...

final BytesReference bytes = serializedState.updateAndGet(tuple -> {
    if (tuple.v1() == state.version) {
        return tuple;
    } else {
        try {
            return new Tuple(state.version(), ClusterStateUtils.serializeClusterState(state, node, true));
        } catch (IOException e) {
            throw new RuntimeException(e);
        }
    }
});
```

I think that works the same as the logic here. Not a huge deal, but it makes the intent a little more clear that you're only ever memoizing a single version of the cluster state.

shwetathareja commented on server/src/internalClusterTest/java/org/opensearch/cluster/coordination/ZenDiscoveryIT.java:
```diff
@@ -117,6 +117,7 @@ public void testHandleNodeJoin_incompatibleClusterState() throws InterruptedExce
 
         final CompletableFuture<Throwable> future = new CompletableFuture<>();
         DiscoveryNode node = state.nodes().getLocalNode();
+
```
[minor]: unintended change.

amkhar replied:
This file is not in diff anymore, identical to main.

amkhar commented on server/src/main/java/org/opensearch/cluster/coordination/ClusterStateUtils.java:
```diff
@@ -0,0 +1,98 @@
+ *
+ * @opensearch.internal
+ */
+public final class ClusterStateUtils {
```
As this class is already doing decompression, deserialization etc, ClusterStateSerialiser doesn't seem right.

We have plans to observe if any other writable object also needs similar compression etc, so I actually wanted to make it more generic(like CompressionHelper) but just named ClusterStateUtils as there are no use case right now(as suggested by Shweta)

I'll keep it as it is for now.

amkhar replied:
Thanks for suggestion.
Won't this `Codec` word will be a bit confusing ? there are Lucene codecs also. Not sure if people will look for this class if they want to use some simple utility methods or want to add something relevant.

andrross commented on server/src/main/java/org/opensearch/cluster/coordination/ClusterStateUtils.java:
```diff
@@ -0,0 +1,98 @@
+     * @return reference to serialized bytes
+     * @throws IOException if writing on the compressed stream is failed.
+     */
+    public static BytesReference serializeClusterState(Writeable writer, DiscoveryNode node, boolean isFullClusterState)
```
A few things are a bit weird about the signature here. It looks like the usage of this method is either:

```
serializeClusterState(ClusterState, node, true);
```
or 

```
serializeClusterState(Diff<ClusterState>, node, false);
```

It would probably be better to use overloading and ditch the boolean, e.g.:

```
serializeClusterState(ClusterState, DiscoveryNode);
serializeClusterState(Diff<ClusterState>, DiscoveryNode);
```

That being said, the bigger issue here is that this is essentially defining a serialization schema (version number, boolean indicating whether it is full or diff, and then the content), but knowledge of how to read this schema is spread to JoinHelper and PublicationTransportHandler. It would be better to keep knowledge of this schema encapsulated in a single class. Is it possible to use a discriminated union pattern here where you have a class defined that can hold either or full cluster state or a diff and then the callers don't have to have knowledge of the serialization schema?

amkhar replied:
That's a nice suggestion @andrross . 

>It would probably be better to use overloading and ditch the boolean, e.g.:
serializeClusterState(ClusterState, DiscoveryNode);
serializeClusterState(Diff<ClusterState>, DiscoveryNode);

The implementation is exactly same for both the cases, keeping two different method looks okay me. Internally, I'll create a single method and call it from both of them. Though a single method where we detect full/diff based on object(instanceof) is good enough. So, I'll incorporate it.

Code would look something like 

```
public static BytesReference serializeClusterState(ClusterState clusterState, DiscoveryNode node) throws IOException {
        return serializeClusterStateInternal(clusterState, node, true);
    }

public static BytesReference serializeClusterState(Diff<ClusterState> clusterStateDiff, DiscoveryNode node) throws IOException {
        return serializeClusterStateInternal(clusterStateDiff, node, false);
    }

private static BytesReference serializeClusterStateInternal(Writeable writer, DiscoveryNode node, boolean isFullClusterState)
        throws IOException {
        final BytesStreamOutput bStream = new BytesStreamOutput();
        try (StreamOutput stream = new OutputStreamStreamOutput(CompressorFactory.COMPRESSOR.threadLocalOutputStream(bStream))) {
            stream.setVersion(node.getVersion());
            stream.writeBoolean(isFullClusterState);
            writer.writeTo(stream);
        }
        final BytesReference serializedByteRef = bStream.bytes();
        logger.trace("serialized writable object for node version [{}] with size [{}]", node.getVersion(), serializedByteRef.length());
        return serializedByteRef;
    }
```

On the other suggestion :
Just so we all are on the same page, here I was trying make code changes in PublicationTransportHandler so there is no code duplication for same logic(state compression/decompression). So I'm not changing the existing code, just moved it to a common place. That's why the utility class was created. 

I hope it's okay to take this (discriminated union pattern one) change separately. I can create a new issue for this cosmetic change.

andrross replied:
> I was trying make code changes in PublicationTransportHandler so there is no code duplication for same logic(state compression/decompression). So I'm not changing the existing code, just moved it to a common place. That's why the utility class was created.

Thanks for the clarification. There was a nuance here that I missed. My hangup is that PublicationTransportHandler is defining a serialization schema with the following logic:

```
stream.setVersion(nodeVersion);
stream.writeBoolean(false);
diff.writeTo(stream);
```

If JoinHandler and PublicationTransportHandler are not exchanging data between each other, then I don't think they should share that _schema_. I think JoinHandler doesn't even need the boolean or the ability to write the diff, and by coupling this there's now an invalid code path in JoinHandler. Is that right?

The other downside here is that if PublicationTransportHandler needs to evolve the data that it writes (by adding a field for example), then this code will have to be untangled or JoinHandler will have to updated to handle the new field, which may or may not make sense.

amkhar replied:
>The other downside here is that if PublicationTransportHandler needs to evolve the data that it writes (by adding a field for example), then this code will have to be untangled or JoinHandler will have to updated to handle the new field

JoinHelper and PTH are not sharing data per se, but they're doing exactly same thing. In both cases, we are sending the cluster state to other node. First we need to remember that the way these variables are written on stream, that order will never change because based on that order only all existing nodes are functional. 
In case we add a new field on PTH (a boolean or int). If it doesn't make sense for JoinHelper, JH reader doesn't need to be updated as it'll function fine. On the other hand if JH requires some new variable, if that intelligence is not required by PTH, that reader will keep functioning fine. 
If some (must know) intelligent variable is added, which can't be avoided for reading cluster state, then both reader part needs to be updated.
I'm still not able to think about any of the use cases. I made this single util class so all code related to compression, decompression, serialization, de-serialization can be put in same place, so whatever is being written on stream is available for all the readers. It's up to the reader if they want to use it or not.

I think we can wait for the time when there is a real use case evolved in this flow, if that really requires a de-coupling, I'll be happy to take it up. 
Note : I'm anyway planning to open up a new issue about this more refactoring. I'll link it here.

shwetathareja replied:
The reason, code was refactored between JoinHelper and PublicationTransportHandler is both publish the cluster state and can use the same logic and optimization in future. PublicationTransportHandler has an optimization to publish diff as well but JoinHandler always publishes the full state. I don't see we publishing diff from JoinHelper so boolean is not needed.

amkhar replied:
@andrross @shwetathareja 

Ack, I understand the concern of sending extra boolean over the stream which will not be utilized even in future. I'll keep the compression/decompression code same and separate out the code writing on the stream.

Will update the pseudo code here before pushing the actual commit, so we're all on the same page.

amkhar replied:
There are 2 ways to achieve this separation. 

**First : Just create a new method in ClusterStateUtils class for node join which doesn't send boolean. So, we can either know about the caller from some ENUM or directly with a different method name. If we choose ENUM, existing method can then have if else etc.**

```
public static BytesReference serializeClusterStateForJoin(Writeable writer, DiscoveryNode node)
        throws IOException {
        final BytesStreamOutput bStream = new BytesStreamOutput();
        try (StreamOutput stream = new OutputStreamStreamOutput(CompressorFactory.COMPRESSOR.threadLocalOutputStream(bStream))) {
            stream.setVersion(node.getVersion());
            writer.writeTo(stream);
        }
        final BytesReference serializedByteRef = bStream.bytes();
        logger.trace("serialized writable object for node version [{}] with size [{}]", node.getVersion(), serializedByteRef.length());
        return serializedByteRef;
    }
```

**Second way would be to create separate classes for these headers which we're adding on top of actual content(cluster state).**

```
public class NodeVersionSchemaBase implements Writeable {

    private final Version nodeVersion;

    public NodeVersionSchemaBase(Version nodeVersion) {
        this.nodeVersion = nodeVersion;
    }

    public Version getNodeVersion() {
        return nodeVersion;
    }

    @Override
    public void writeTo(StreamOutput out) throws IOException {
        out.setVersion(nodeVersion);
    }
}


public class ClusterStateUpdateSchema extends NodeVersionSchemaBase implements Writeable {
    private final boolean isFullClusterState;

    public ClusterStateUpdateSchema(Version nodeVersion, boolean isFullClusterState) {
        super(nodeVersion);
        this.isFullClusterState = isFullClusterState;
    }

    @Override
    public void writeTo(StreamOutput out) throws IOException {
        super.writeTo(out);
        out.writeBoolean(isFullClusterState);
    }
}

```


Based on the schema our Util class code will be changed something like 
```
 public static BytesReference serializeClusterState(ClusterState clusterState, DiscoveryNode node) throws IOException {
        return serializeClusterStateOrDiff(clusterState, new ClusterStateUpdateSchema(node.getVersion(), true));
    }

public static BytesReference serializeClusterState(Diff<ClusterState> clusterStateDiff, DiscoveryNode node) throws IOException {
        return serializeClusterStateOrDiff(clusterStateDiff, new ClusterStateUpdateSchema(node.getVersion(), false));
    }

public static BytesReference serializeClusterStateForJoin(Writeable writer, DiscoveryNode node)
        throws IOException {
        return serializeClusterStateOrDiff(writer, new NodeVersionSchemaBase(node.getVersion()));
    }
```

And finally the end method will look something like 

```
private static BytesReference serializeClusterStateOrDiff(Writeable writer,
                                                              NodeVersionSchemaBase streamHeaders) throws IOException {
        final BytesStreamOutput bStream = new BytesStreamOutput();
        try (StreamOutput stream = new OutputStreamStreamOutput(CompressorFactory.COMPRESSOR.threadLocalOutputStream(bStream))) {
            streamHeaders.writeTo(stream);
            writer.writeTo(stream);
        }
        final BytesReference serializedByteRef = bStream.bytes();
        logger.trace("serialized writable object for node version [{}] with size [{}]", streamHeaders.getNodeVersion(),
            serializedByteRef.length());
        return serializedByteRef;
    }
```

Again, here I'm using Util class itself to create these schema objects as currently we've already moved that complexity from JoinHelper as well as PTH. If needed this can be directly created from both of those 2 classes, as they are the callers and it can be their responsibility. Right now, it's not a big schema, so keeping it here is also fine from my side. As long as we keep separate methods for separate callers like I created a new method for Node Join.

Extendibility support :

2nd approach allows us to modify the headers for one caller without touching anything for second caller and common method doesn't need any changes. Only schema classes would require changes, and schema creation code.

Please suggest if 2nd approach looks good, or what more improvements can be done there ?
**Note 1** : In case JoinHelper needs to add a different variable in stream, we'll create a new class similar to `ClusterStateUpdateSchema` may be `ValidateJoinStateSchema` and add data there. This NodeVersionBaseSchema is added so it's always added to the stream (else serialization fails for ClusterState).
**Note 2**: I thought of using composition here also instead of inheritance, but wanted to keep the end method with same type(NodeVersionSchemaBase streamHeaders, so any caller can pass their headers as they all will be inherited from base class) for stream headers, so used inheritance to avoid any if else (or more methods).

andrross replied:
I was thinking of something a little simpler. Essentially replace `ClusterStateUtils` with this class:

```java
public final class CompressedStreamUtils {
    public static BytesReference createCompressedStream(Version version, CheckedConsumer<StreamOutput, IOException> outputConsumer) throws IOException {
        final BytesStreamOutput bStream = new BytesStreamOutput();
        try (StreamOutput stream = new OutputStreamStreamOutput(CompressorFactory.COMPRESSOR.threadLocalOutputStream(bStream))) {
            stream.setVersion(version);
            outputConsumer.accept(stream);
        }
        return bStream.bytes();
    }

    public static StreamInput decompressBytes(BytesTransportRequest request, NamedWriteableRegistry namedWriteableRegistry)
        throws IOException {
        final Compressor compressor = CompressorFactory.compressor(request.bytes());
        final StreamInput in;
        if (compressor != null) {
            in = new InputStreamStreamInput(compressor.threadLocalInputStream(request.bytes().streamInput()));
        } else {
            in = request.bytes().streamInput();
        }
        in.setVersion(request.version());
        return new NamedWriteableAwareStreamInput(in, namedWriteableRegistry);
    }
}
```

Both PTH and JoinHelper can use those functions and retain full control over what they read and write from the stream. It seems like this should work? I did a [quick prototype here](https://github.com/andrross/OpenSearch/commit/ea757e21d665871afa33ccc7fabd67cdfe2cee4a) (didn't update the tests though).

amkhar replied:
@andrross  

Sure I saw the prototype code. Only difference I see is that we're writing boolean from PTH class itself. 

I suggested this also in my previous [comment](https://github.com/opensearch-project/OpenSearch/pull/7321#discussion_r1214268115) 

>Again, here I'm using Util class itself to create these schema objects as currently we've already moved that complexity from JoinHelper as well as PTH. If needed this can be directly created from both of those 2 classes, as they are the callers and it can be their responsibility.

So we can create those - `new ClusterStateUpdateSchema(node.getVersion(), false)` from PTH class and pass it in the util method  -> 

`public static BytesReference serializeClusterStateOrDiff(Writeable writer, NodeVersionSchemaBase streamHeaders)`

I hope that is similar to what you're suggesting :) . I'm okay if you want that in `CheckedConsumer<StreamOutput, IOException> outputConsumer` format.

@shwetathareja  @andrross  
Let's finalize on one coding approach. Thanks Andrew for the prototype code.

@shwetathareja
Let me know what do you prefer. I'll make code changes accordingly.

shwetathareja replied:
@amkhar Lets keep the code simple and let the callers PTH and JoinHelper pass the supplier to handle the serialization of their objects (with/ without boolean). We don't need new interface and classes.

amkhar replied:
Thank you both.

@andrross I am planning to take your code and push a new commit out of it.

Please explain if there is any other change which is required after this ?

I want to incorporate all required changes in one go.

andrross replied:
No other comments from my side. Thanks!

amkhar replied:
Thanks @andrross .

amkhar commented:
> SearchWeightedRoutingIT

Flaky test : https://github.com/opensearch-project/OpenSearch/issues/5957

amkhar commented:
>org.opensearch.search.backpressure.SearchBackpressureIT.testSearchShardTaskCancellationWithHighCpu

Flaky test link : https://github.com/opensearch-project/OpenSearch/issues/7750

>org.opensearch.cluster.allocation.AwarenessAllocationIT.testThreeZoneOneReplicaWithForceZoneValueAndLoadAwareness

Flaky test link : https://github.com/opensearch-project/OpenSearch/issues/3603

andrross commented on server/src/main/java/org/opensearch/cluster/coordination/JoinHelper.java:
```diff
@@ -208,22 +218,53 @@ public ClusterTasksResult<JoinTaskExecutor.Task> execute(ClusterState currentSta
+        BytesTransportRequest request
+    ) throws IOException {
+        try (StreamInput input = CompressedStreamUtils.decompressBytes(request, namedWriteableRegistry)) {
+            input.setVersion(request.version());
```
```suggestion
```

This line is not needed because the helper sets the version, right?

andrross reviewed (approved):
Looks good, just one super minor comment. Thanks for working through this and dealing with all my comments!
